admin/acceptance.py

```diff
@@ -1289,24 +1291,31 @@ def capture_upstart(reactor, host, output_file):
             (b"flocker", b"flocker-container-agent"),
             (b"flocker", b"flocker-docker-plugin"),
             (b"upstart", b"docker")]:
-        path = FilePath(b'/var/log/').child(directory).child(service + b'.log')
-        formatter = TailFormatter(output_file, host, service)
-        ran = run_ssh(
-            reactor=reactor,
-            host=host,
-            username='root',
-            command=[
-                b'tail',
-                b'-F',
-                path.path
-            ],
-            handle_stdout=formatter.handle_output_line,
+
+        def pull_logs_for_process(directory=directory, service=service):
+            path = FilePath(
+                b'/var/log/').child(directory).child(service + b'.log')
+            formatter = TailFormatter(output_file, host, service)
+            ran = run_ssh(
+                reactor=reactor,
+                host=host,
+                username='root',
+                command=[
+                    b'tail',
+                    b'-F',
+                    path.path
+                ],
+                handle_stdout=formatter.handle_output_line,
+            )
+            ran.addErrback(write_failure, logger=None)
+            # Deliver a final empty line to process the last message
+            ran.addCallback(lambda ignored, formatter=formatter:
+                            formatter.handle_output_line(b""))
+            return ran
+
```

Comment:
Note added.

Yea the trouble with getting logs at the end is that if the VM randomly gets killed in a way that we can't bring it back we can't get the logs off of it.

That said... I don't know of any flocker failure modes that kill the VM, so maybe this is over-engineered. But in general, I prefer streaming logs over hoping you can retrieve them later.